orchard_group_hash.py

```diff
@@ -0,0 +1,168 @@
+
+        b.append(bi_ctx.digest())
+        assert len(b[i]) == b_in_bytes
+
+    return b''.join(b[1:])[0:len_in_bytes]
+
+def hash_to_field(msg, dst):
+    k = 256
+    count = 2
+    m = 1
+
+    L = cldiv(math.ceil(math.log2(p)) + k, 8)
+    assert L == 512/8
+
+    len_in_bytes = count * m * L
+    uniform_bytes = expand_message_xmd(msg, dst, len_in_bytes)
+
+    elements = []
+    for i in range(0, count):
+        for j in range(0, m):
+            elm_offset = L * (j + i * m)
+            tv = uniform_bytes[elm_offset:elm_offset+L]
+            elements.append(Fp(beos2ip(tv), False))
+
+    assert len(elements) == count
+
+    return elements
+
+def map_to_curve_simple_swu(u):
+    # The notation below follows Appendix F.2 of the Internet Draft
+    zero = Fp(0)
+    assert zero.inv() == Fp(0)
+
+    A = PALLAS_ISO_A
+    B = PALLAS_ISO_B
+    Z = Fp(-13, False)
+    c1 = -B / A
+    c2 = Fp(-1) / Z
+
+    tv1 = Z * u.exp(2)
+    tv2 = tv1.exp(2)
+    x1 = tv1 + tv2
+
+    x1 = x1.inv()
+    e1 = x1 == Fp(0)
+    x1 = x1 + Fp(1)
+
+    if e1:
+        x1 = c2
+    else:
+        x1 = x1
+
+    x1 = x1 * c1      # x1 = (-B / A) * (1 + (1 / (Z^2 * u^4 + Z * u^2)))
+    gx1 = x1.exp(2)
+    gx1 = gx1 + A
+    gx1 = gx1 * x1
+    gx1 = gx1 + B             # gx1 = g(x1) = x1^3 + A * x1 + B
+    x2 = tv1 * x1            # x2 = Z * u^2 * x1
+    tv2 = tv1 * tv2
+    gx2 = gx1 * tv2           # gx2 = (Z * u^2)^3 * gx1
+
+    e2 = (gx1.sqrt() is not None)
+
+    x = x1 if e2 else x2    # If is_square(gx1), x = x1, else x = x2
+    yy = gx1 if e2 else gx2  # If is_square(gx1), yy = gx1, else yy = gx2
+    y = yy.sqrt()
+
+    e3 = u.sgn0() == y.sgn0()
+
+    y = y if e3 else -y  #y = CMOV(-y, y, e3)
+
+    return orchard_iso_pallas.Point(x, y)
+
+def group_hash(d, m):
+    dst = d + b"-" + b"pallas" + b"_XMD:BLAKE2b_SSWU_RO_"
+
+    elems = hash_to_field(m, dst)
+    assert len(elems) == 2
+
+    q = [map_to_curve_simple_swu(elems[0]).iso_map(), map_to_curve_simple_swu(elems[1]).iso_map()]
+
+    return q[0] + q[1]
+
+
+def main():
+    fixed_test_vectors = [
+        # This is the Pallas test vector from the Sage and Rust code (in affine coordinates).
+        (b"z.cash:test", b"Trans rights now!", Point(Fp(10899331951394555178876036573383466686793225972744812919361819919497009261523),
+                                                     Fp(851679174277466283220362715537906858808436854303373129825287392516025427980))),
+    ]
+
+    for (domain, msg, point) in fixed_test_vectors:
+        gh = group_hash(domain, msg)
+        assert gh == point
+
+    test_vectors = [(domain, msg) for (domain, msg, _) in fixed_test_vectors]
+
+    render_tv(
+        render_args(),
+        'orchard_group_hash',
+        (
+            ('domain', 'Vec<u8>'),
+            ('msg', 'Vec<u8>'),
+            ('point', '[u8; 32]'),
+        ),
+        [{
+            'domain': domain,
+            'msg': msg,
+            'point': bytes(group_hash(domain, msg)),
```

Comment:
Why recompute `group_hash` here when we already checked it is equal to the point in `fixed_test_vectors`?

Reply:
Never mind, I see why now that you've added the random generation code.